### PR TITLE
perf(drive)!: verify instant lock signatures with Drive

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

Core RPC was never designed for public access and according to our tests is a weak point in the system. Under the load, instant lock signature verification takes seconds. Since users can trigger signature verification via invalid IdentityCreateTransition for free, this open a DoS attack vector. 

## What was done?

- Own implementation of Instant Lock signature verification for InstantAssetLockProof without calling Core RPC
- Introduced support for DIP24 rotating quorums
- Base config network in dashmate changed to mainnet
- Updated quorums configuration for local, testnet and mainnet
- Added missing logging for invalid asset lock based state transitions
- Updated invalid signature error for better verbosity
- Fixed internal error if chain lock signature is not valid

## How Has This Been Tested?

- Unit tests skip instant lock signature verification with `disable_instant_lock_signature_verification` test config option.
- Implemented Instant Lock quorums and signing in strategy tests
- Updated `run_chain_top_up_identities` to use InstantLock singing with DIP24 rotating quorums
- Updated `run_chain_insert_one_new_identity_per_block_with_block_signing` to use InstantLock singing with DIP8 classic quorums
- With test suite against local network
- With test suite against a devnet

## Breaking Changes

InstantAssetLockProof might give different validation results, since our implementation supports only recent signatures, compared with Core RPC. This means the previous blockchain data might become invalid.
 

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs [hunk 1]
```diff
@@ -45,12 +44,29 @@ impl AssetLockProofValidation for InstantAssetLockProof {
...
+            // We should be able to disable instant lock versification for integration tests
+            if platform_ref
+                .config
+                .testing_configs
+                .disable_instant_lock_signature_verification
+            {
+                return Ok(validation_result);
+            }
+
+            // TODO: Shouldn't we add an operation for fees?
```
Comment (shumkov) [inline packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs:56]: Shouldn't we add an operation to calculate fees?

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs [hunk 2]
```diff
@@ -133,6 +171,10 @@ where
         let new_validator_sets = quorum_infos
             .into_iter()
             .map(|(quorum_hash, info_result)| {
+                // TODO: There is a non-determinism here because quorum info is not based on core chain locked height
```
Comment (shumkov) [inline packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs:174]: Non-determinism

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/choose_quorum/mod.rs [hunk 3]
```diff
@@ -21,7 +21,7 @@ where
     C: CoreRPCLike,
 {
     /// Based on DIP8 deterministically chooses a pseudorandom quorum from the list of quorums
-    ///
+    // TODO: use CoreQuorumSet.select_quorums instead
```
Comment (shumkov) [inline packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/choose_quorum/mod.rs:24]: Shall we remove and use new Quorum structures for chain lock in tests as well?

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs [hunk 4]
```diff
@@ -63,9 +68,11 @@ where
             return Ok(None); // the chain lock is too far in the future or the past to verify locally
         }
 
-        let mut selected_quorum_sets = platform_state
+        let mut selected_quorums = platform_state
             .chain_lock_validating_quorums()
-            .select_quorums(chain_lock_height, verification_height);
+            .select_quorums_for_heights(chain_lock_height, verification_height);
+
+        // TODO: We can use chain_lock.request_id()
```
Comment (shumkov) [inline packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs:75]: Shall we?

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs [hunk 5]
```diff
@@ -102,20 +140,26 @@ where
...
+                    &key,
                     None,
                 )?;
 
-                Ok((*key, quorum_info_result))
+                // In case if current chain locked height so far in the past,
+                // long enough for a new quorum to be formed, quorum info (which is not
+                // based on chain locked height) will respond with altered quorum index,
+                // so we need to use quorum index from the deterministic `quorum listextended` response
+                quorum_info_result.quorum_index = details.quorum_index.unwrap_or(0);
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs:160]: take this out

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs [hunk 6]
```diff
@@ -63,9 +68,11 @@ where
             return Ok(None); // the chain lock is too far in the future or the past to verify locally
         }
 
-        let mut selected_quorum_sets = platform_state
+        let mut selected_quorums = platform_state
             .chain_lock_validating_quorums()
-            .select_quorums(chain_lock_height, verification_height);
+            .select_quorums_for_heights(chain_lock_height, verification_height);
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs:73]: rename back to select_quorums

File: packages/rs-drive-abci/src/execution/platform_events/mod.rs [hunk 7]
```diff
@@ -26,3 +26,6 @@ pub(in crate::execution) mod block_start;
 
 /// Verify the chain lock
 pub(in crate::execution) mod core_chain_lock;
+
+/// Instant lock methods
+pub(in crate::execution) mod core_instant_lock;
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/mod.rs:31]: core_instant_lock -> core_instant_send_lock

File: packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs [hunk 8]
```diff
@@ -45,12 +44,28 @@ impl AssetLockProofValidation for InstantAssetLockProof {
...
-                platform_ref.state.last_committed_core_height(),
-            )?;
+            // We should be able to disable instant lock versification for integration tests
+            if platform_ref
+                .config
+                .testing_configs
+                .disable_instant_lock_signature_verification
+            {
+                return Ok(validation_result);
+            }
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs:54]: add todo to not expose testing_configs in production code

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs (hunk 1 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs:56]: Actually, there are more places in asset lock verification is not covered with fees. I will do it as a separate PR.

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/choose_quorum/mod.rs (hunk 3 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/choose_quorum/mod.rs:24]: Not today. Decided to do it for next versions.

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs (hunk 4 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs:75]: Refactoring for next version

File: packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs (hunk 8 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/validation/state_transition/common/asset_lock/proof/validate/instant/mod.rs:55]: Added feature flag

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs [hunk 9]
```diff
@@ -173,139 +203,214 @@ where
...
+        // Update Instant Lock quorums
+
+        self.update_quorums(
+            QuorumSetType::InstantLock(instant_lock_quorum_type),
+            block_platform_state,
+            platform_state,
+            &extended_quorum_list,
+            is_validator_set_updated,
+            core_block_height,
+        )?;
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs:226]: I would do this only if instant_lock_quorum_type is different to chain_lock_quorum_type right?

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs [hunk 10]
```diff
@@ -3,17 +3,46 @@ use crate::error::Error;
...
 use crate::platform_types::platform_state::v0::PlatformStateV0Methods;
 use crate::platform_types::platform_state::PlatformState;
-use crate::platform_types::signature_verification_quorums::SignatureVerificationQuorumsV0Methods;
+use dashcore_rpc::json::{ExtendedQuorumListResult, QuorumType};
 use std::collections::BTreeMap;
+use std::fmt::Display;
 
-use crate::platform_types::validator_set::v0::{ValidatorSetV0, ValidatorSetV0Getters};
+use crate::platform_types::validator_set::v0::{
+    ValidatorSetV0, ValidatorSetV0Getters, ValidatorSetV0Setters,
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs:11]: unused

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs [hunk 11]
```diff
@@ -150,14 +152,19 @@ where
 
                 let mut engine = sha256d::Hash::engine();
 
-                engine.input(&[self.config.chain_lock_quorum_type() as u8]);
-                engine.input(quorum_hash.as_slice());
+                let mut reversed_quorum_hash = quorum_hash.to_byte_array().to_vec();
+                reversed_quorum_hash.reverse();
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs:156]: why is this now reversed?

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs (hunk 11 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/core_chain_lock/verify_chain_lock_locally/v0/mod.rs:156]: We have done it before as well but inside choose_quorum function. Now it returns QuorumHash, Quorum for consitency (as all other functions there).

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs (hunk 9 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs]: Fixed

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs (hunk 10 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_quorum_info/v0/mod.rs]: Fixed

**Comment by shumkov** [comment]
Agreed with @QuantumExplorer that I will merge this PR and he will do a post review of the recent two changes later.

**Pull request merged**